autopilot/prefattach.go

```diff
@@ -257,3 +257,120 @@ func (p *ConstrainedPrefAttachment) Select(self *btcec.PublicKey, g ChannelGraph
+// joining the network with an amount of available funds to be allocated to
+// channels. Specifically, we consider the degree of each node (and the flow
+// in/out of the node available via its open channels) and utilize the
+// Barabási–Albert model to drive our recommended attachment heuristics. If
+// implemented globally for each new participant, this results in a channel
+// graph that is scale-free and follows a power law distribution with k=-3.
+//
+// The returned scores will be in the range [0.0, 1.0], where higher scores are
+// given to nodes already having high connectivity in the graph.
+//
+// NOTE: This is a part of the AttachmentHeuristic interface.
+func (p *ConstrainedPrefAttachment) NodeScores(g ChannelGraph, chans []Channel,
+	fundsAvailable btcutil.Amount, nodes map[NodeID]struct{}) (
+	map[NodeID]*AttachmentDirective, error) {
+
+	// Fill the map of candidates to return.
+	// NOTE: we'll set the addrs, score and chansize below.
+	candidates := make(map[NodeID]*AttachmentDirective)
+	for nID := range nodes {
+		pub, err := btcec.ParsePubKey(nID[:], btcec.S256())
+		if err != nil {
+			return nil, err
+		}
+		candidates[nID] = &AttachmentDirective{
+			NodeKey: &btcec.PublicKey{
+				X: pub.X,
+				Y: pub.Y,
+			},
+			NodeID: nID,
+		}
+	}
+
+	// Count the number of channels in the graph. We'll also count the
+	// number of channels as we go for the nodes we are interested in.
+	var graphChans int
+	nodeChanNum := make(map[NodeID]int)
+	if err := g.ForEachNode(func(n Node) error {
+		var nodeChans int
+		err := n.ForEachChannel(func(_ ChannelEdge) error {
+			nodeChans++
+			graphChans++
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		// If this node is not among our nodes to score, we can return
+		// early.
+		nID := NodeID(n.PubKey())
+		if _, ok := nodes[nID]; !ok {
+			return nil
+		}
+
+		// Otherwise we'll record the number of channels, and also
+		// populate the address in our channel candidates map.
+		nodeChanNum[nID] = nodeChans
+		candidates[nID].Addrs = n.Addrs()
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	existingPeers := make(map[NodeID]struct{})
+	for _, c := range chans {
+		existingPeers[c.Node] = struct{}{}
+	}
+
+	// For each node in the set of nodes, count their fraction of channels
+	// in the graph, and use that as the score.
+	for nID, nodeChans := range nodeChanNum {
+		// As channel size we'll use the maximum channel size available.
+		chanSize := p.constraints.MaxChanSize
+		if fundsAvailable-chanSize < 0 {
+			chanSize = fundsAvailable
+		}
+
+		_, ok := existingPeers[nID]
+
+		var score float64
+		switch {
+
+		// If the node is among or existing channel peers we give it a
+		// score of 0, as we don't need another channel.
+		case ok:
+			score = 0
+
+		// If there are no channels in the graph we cannot determine
+		// any preferences, so we set all scores to zero.
+		case graphChans == 0:
+			score = 0
+
+		// Set score 0 if channel is too small.
```

Comment:
Seems we can just check this above (outside of the select loop), and continue altogether, not even including this node in the final scoring map.

Reply:
Done.